api/src/main/java/org/eclipse/microprofile/config/Config.java

```diff
@@ -82,55 +95,245 @@
  * @author <a href="mailto:emijiang@uk.ibm.com">Emily Jiang</a>
  * @author <a href="mailto:gunnar@hibernate.org">Gunnar Morling</a>
- *
+ * @author <a href="mailto:david.lloyd@redhat.com">David M. Lloyd</a>
  */
 @org.osgi.annotation.versioning.ProviderType
 public interface Config {
 
     /**
      * Return the resolved property value with the specified type for the
      * specified property name from the underlying {@link ConfigSource ConfigSources}.
-     * 
-     * If this method gets used very often then consider to locally store the configured value.
-     *
-     * @param <T>
-     *             The property type
-     * @param propertyName
-     *             The configuration propertyName.
-     * @param propertyType
-     *             The type into which the resolve property value should get converted
-     * @return the resolved property value as an object of the requested type.
-     * @throws java.lang.IllegalArgumentException if the property cannot be converted to the specified type.
-     * @throws java.util.NoSuchElementException if the property isn't present in the configuration.
+     * <p>
+     * If this method gets used very often, consider locally storing the configured value.
+     *
+     * @param <T> the result type
+     * @param propertyName the configuration property name (must not be {@code null})
+     * @param propertyType the type into which the resolved property value should be converted (must not be {@code null})
+     * @return the resolved property value as an object of the requested type
+     * @throws java.lang.IllegalArgumentException if the property cannot be converted to the specified type
+     * @throws java.util.NoSuchElementException if the property isn't present in the configuration, or is present
+     *  but has an empty value
      */
     <T> T getValue(String propertyName, Class<T> propertyType);
 
     /**
      * Return the resolved property value with the specified type for the
+     * specified property name from the underlying {@link ConfigSource ConfigSources}, using the given
+     * default value if the property is not found.
+     * <p>
+     * If this method gets used very often, consider locally storing the configured value.
+     *
+     * @param <T> the result type
+     * @param propertyName the configuration property name (must not be {@code null})
+     * @param propertyType the type into which the resolved property value should be converted (must not be {@code null})
+     * @param defaultValue the default value to return if no property with the given name is found
+     * @return the resolved property value as an object of the requested type
+     * @throws java.lang.IllegalArgumentException if the property cannot be converted to the specified type
+     * @throws java.util.NoSuchElementException if the property has an empty value
+     */
+    <T> T getValueWithDefault(String propertyName, Class<T> propertyType, T defaultValue);
```

Comment:
It cannot be removed.  `NoSuchElementException` does not mean not-present, it means empty.  We need some exception to represent this case, and this is the exception used already today (with 1.3 and earlier); changing that would be drastically incompatible.  Whereas with this behavior almost every existing case will continue to work the same.  I recognize it is not ideal, but if one just reads the documentation it should be clear what it means: it's thrown when the property has an empty value.

The property type has to be given because the default value can be any subtype, including null which is a subtype of everything.  It would be possible to extrapolate the type in some cases but then the user might end up with a different converter than the one they were wanting.

Reply:
> `NoSuchElementException` does not mean not-present, it means empty. We need some exception to represent this case, and this is the exception used already today (with 1.3 and earlier); changing that would be drastically incompatible.

I was trying to point out that you don't need to raise the NoSuchEelementException from this method at all.  It's a newly introduced method by this pull, so there is no incompatibility with existing usage.  Given that the point of the method is to let the user provide a default value, there should always be a value to return (absent a conversion error, which you already cover with IllegalArgumentException) - either an explicitly configured value that evaluates to something (non-empty) or otherwise the default.  I thought this was the behavior you were wanting to achieve with "" and other expressions that evaluate to no value specified.

What I've been trying to emphasize as an important design point is that a single configuration property value, given a specific type, ought to have a single meaning.   The conclusion that I thought we had reached was that the zero-length string (""), and by extension other values that evaluate to 'no value specified', would have the meaning of unconfigure.  If that is what we are presenting to the user, then the expected behavior for a method to which you supply a default would be to use the default when no value is configured and should be identical to what you get if the value is unconfigured.

Reply:
But then you re-introduce the problem that you can't override a default value with an empty value.  So if I have a property which, for example, represents a list of contexts to propagate, and has a default value of `"foo,bar"`, but the user wants to explicitly propagate no contexts, the user could specify `""` and reasonably expect no contexts to propagate.  But if, as you propose, either a non-empty value or the default value is returned, the result would be `"foo,bar"`, invalidating the purpose of the change; the user would not be able to specify "no contexts" without a special sentinel value, which doesn't work for all types and thus would require custom converters in some cases but not others.

Reply:
Yes, I understand that, that is the exact the scenario that I need to preserve for MicroProfile Context Propagation.  However, I'm conceding that we need to sacrifice that because otherwise MicroProfile Config becomes too complex and doesn't make any sense to the average user.  We will have to deal with it, either by introducing our own constant for empty value/list, or convincing MP Config to standardize one.

Reply:
I really don't think the change can be done without having a first-order concept of an "empty value", and a first-order concept of a "missing key", and a first-order concept of a "default value", and defining behaviors in those terms:

* An *empty value* is a value that would convert to nothing (for example, imagine a property of type `Double` that was given as an empty string); empty values can be given as explicit values for configuration keys
* A *default value* is a value for a configuration key that implicitly exists in a "virtual" lowest-priority configuration source, i.e., it's only used when no other configuration source gave a value
* A *missing key* is a configuration key that is not specified in any config source (including as a default value); in absence of a *default value*, *missing keys* resolve to *empty values*

Since we don't want to force every type to have an "empty" representation - especially because some simply can't - one might consider to use `null` to mean "empty" for all types; however, regardless of one's feelings about `null`, `getValue` is not expected to return it historically so that would be a major breaking change, and in fact is the exact reason that `getOptionalValue` exists, returning (as it does) its `null`-wrapping payload.

But because of this, `Optional` simply cannot represent "missing", because if it did, what would represent "empty"?  Sure we could add an `Emptiable<T>` and then have an `Optional<Emptiable<T>>` variation, but isn't that worse?  Now the user has to figure out what to do in the empty case, or the missing case, or the missing-and-empty case, and what would `Emptiable<Optional<T>>` mean anyway?  Or `Emptiable<String>`, or `Emptiable<OptionalInt>`?  You'd end up with as many rules as you have permutations of types, values, and converters.

At least the rules as proposed are consistent and clear, if one reads the docs.  They're even intuitive, especially once you consider questions like:

* What do I want to get back if I call `getValue` and pass in a `Converter<OptionalInt>` and the value is missing?  Or explicitly empty?  What if the value defaults to "10" but can also be explicitly cleared?  Shouldn't the result be `OptionalInt.empty()`?  How does that compare to if I pass in a `Converter<Optional<Integer>>`?  Shouldn't they behave the same i.e. get back an `Optional.empty()`?
* Compare to this case: what do I want to get if I call `getValue` with a `Converter<String>` and the value is empty?  An empty string?
* But then... if I call `getValue` with a `Converter<List<String>>`... now I come back to the janky list case where I get a list of blanks (of possibly unexpected size) in some cases.  And then, using the same logic, a `Converter<List<Double>>` gets... an exception!  Why?  Because we don't represent empty specially, so the conversion fails for lists such as `"123.0,${foo:},456.0"` in the absence of a `foo` property, because `Double` has no empty representation and so cannot be elided from the list.

Without the ability to represent empty specially, generic enclosing converters (like those needed to return arrays or lists of things, or optional-ize them, etc.) become impossible because they can never, ever consistently handle empties, because empty handling varies by type.  Therefore we have a special representation of emptiness: at a converter level, this is `null`.  This allows a generic list converter to delegate conversion of its elements, removing empty segments and producing an expected result. But now we have a new problem: what if the list ends up empty?  Should we return an empty list?  No!  Because now the list converter in turn cannot be encapsulated, closing the door on more complex configuration value types.  Instead, the empty list returns `null` as well, as must *every other* converter type that would produce an empty value, allowing converters to nest arbitrarily.

In other words, types that have special empty values cannot be nested within other types.  But `Optional` serves that purpose perfectly: if the value is empty, then you can use `orElse` to use whatever empty representation you like, be it an empty list, empty string, etc.

It all makes sense once you accept the above three definitions.  Then you know that every key has a value - even if it's an *empty value* - and `getOptionalValue` lets you return an empty value but `getValue` does not, returning `NoSuchElementException` in this case.  Since empty values behave identically to pre-1.4 missing values, it makes sense to continue the same semantics moving forward.  In addition, by specifying a custom converter, customized empty values can also be used, though in most cases `Optional` is sufficient.  And by adding a way to specify a use-site default, one can conveniently get a value that is used when the key is *missing*.

Reply:
While I wouldn't have defined default value as you do, I can see how doing so is helpful to the solution you are aiming for.  Otherwise, I can follow the logic up until,

> But because of this, `Optional` simply cannot represent "missing", because if it did, what would represent "empty"?

This is where I get lost in the logic, because, as someone who is more familiar with java.util.Optional than MP Config, Optional does have a natural representation of "missing", which is `Optional.empty()`.  From previous comments, I know you don't agree with that and I don't want to start the argument again.  But if you assume for a moment, despite the unfortunate overloading of terminology, that `Optional.empty()` is the natural way of representing "missing/not present", that leaves the question of how does an empty value get represented, which would be "" for String, size 0 array for arrays, and a failure to create the Optional at all if not representable.

We already decided, I wouldn't push that approach any further.  I'll certainly acknowledge that your solution offers more functionality and flexibility, and is very creative in achieving that.  I don't see it as intuitive, though, and have tried to explore some ways of improving on that.  Unfortunately, it seems the conclusion we are reaching is that doing so would start to limit the function you are aiming to achieve, and so isn't acceptable.

At this point, I would highly recommend that this should be reviewed by MP Config contributors/experts, rather than by myself, who is just a user of MP Config.  What you have proposed may very well be the right strategic direction for MP Config to take.  I'm not the right person to make that sort of judgement.  All of the feedback that I have to offer, I have already offered.  You and the other contributors to MP Config should just make the decision.